lib/Doctrine/DBAL/Driver/PDOPgSql/Driver.php

```diff
@@ -59,7 +59,7 @@ public function connect(array $params, $username = null, $password = null, array
              * - passing client_encoding via the 'options' param breaks pgbouncer support
              */
             if (isset($params['charset'])) {
-              $pdo->query('SET NAMES \''.$params['charset'].'\'');
+              $pdo->exec('SET NAMES \''.$params['charset'].'\'');
```

Comment:
Careful with `exec` vs `query` - we had issues specifically around switching these in the past, and these changes shouldn't be performed without extreme care here

Reply:
Interesting, can you share more insight please (so I could document it in the inspection documentation)?. I'll also revert this specific case.

Reply:
I can't find an issue right now, but we had problems with multiple drivers in switching between `query` and `exec` for non-SQL stuff such as stored procedures and similar, which do not necessarily use the same underlying protocol for talking between the extension and the SQL server you are connected to.

@morozov can probably say more about that

Reply:
@Ocramius is probably talking about #2752. Also, for the reference, in MySQL listener, we do:

https://github.com/doctrine/dbal/blob/80f42dc7afe91781e698e63b7d362b2f51986935/lib/Doctrine/DBAL/Event/Listeners/MysqlSessionInit.php#L70

which internally calls:

https://github.com/doctrine/dbal/blob/752f9f5c89754bbad04a13ffc0877960658a68ae/lib/Doctrine/DBAL/Connection.php#L1088

With that in mind, I think the change should be safe.

Reply:
Thank you for sharing @morozov!

Reply:
Um… @kalessil I see you reverted it. My point was that changing `query()` to `exec()` was safe since we do the same for MySQL. BTW what kind of inspection found this issue? Was it something like unused function result?

Reply:
Oh, will re-introduce the change then. Nope: it's a specialized inspection from this plugin: https://plugins.jetbrains.com/plugin/7622-php-inspections-ea-extended- (my project)

Reply:
Yeah, I know about Php Inspections, I was just curious which exact rule from the plugin flagged the issue.

Reply:
PHP | Php Inspections (EA Extended) | Control Flow | PDO API usage

Reply:
OK, with MySQL the changing from `query()` to `exec()` can be safe, but must not. PDO maybe is per default not using native prepared statements (https://stackoverflow.com/questions/10113562/pdo-mysql-use-pdoattr-emulate-prepares-or-not). For MSSQL native prepared statements was activated.

As @morozov said, when calling `PDO::query()` without parameters it's not the same as `PDO::exec()`. I've made a test-script:

    $pdo->exec('SET DATEFIRST 1;');
    $pdo->query('SET DATEFIRST 1;');

When looking into `sys.dm_exec_cached_plans` I find to plans for the two calls:


    ----------------------------
    objtype   | text          
    ----------------------------
    Adhoc     | SET DATEFIRST 1;
    Prepared  | SET DATEFIRST 1;

`Adhoc` queries end up in the standard environment while prepared statements are executed in a separate environment. So if you `SET OPTIONS` with prepared statement the next statement will not have access to the new option value.

@kalessil But your inspection gave me the hint what I did wrong, however, it is not generally applicable.

https://github.com/microsoft/msphpsql/issues/385
https://aoeex.com/phile/working-around-scope-identity-not-working-with-pdo
https://docs.microsoft.com/en-us/sql/connect/php/direct-statement-execution-prepared-statement-execution-pdo-sqlsrv-driver?view=sql-server-ver15

Reply:
@algorhythm, thanks for the ping. Re-visiting the discussion, we should drop query->exec pattern.

//cc @ea-inspections-team

Reply:
Thanks for the ping @kalessil ! Done, we've dropped the pattern detection.